In [ ]:
#1. Nhóm khách hàng Best Customers có đặc điểm như thế nào? Số lượng? (1.5đ)
#2. Đâu là các khách hàng rời bỏ?  Số lượng? (1.5đ)
#3. Nhóm khách hàng có khả năng Churned có đặc điểm như thế nào ?  (1.5đ)

SyntaxError: invalid decimal literal (<ipython-input-94-de57c077c4d2>, line 1)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

In [ ]:
import pandas as pd

# Function to get direct download link from Google Drive shared link
def get_direct_link(file_id):
    return f'https://drive.google.com/uc?id={file_id}'


# Đọc dữ liệu từ các tệp Excel
customers_df= pd.read_csv('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_customers_dataset.csv')
geolocation_df= pd.read_csv('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_geolocation_dataset.csv')
items_df= pd.read_csv('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_order_items_dataset.csv')
payments_df= pd.read_csv(('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_order_payments_dataset.csv'))
reviews_df= pd.read_csv(('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_order_reviews_dataset.csv'))
orders_df= pd.read_csv(('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_orders_dataset.csv'))
products_df= pd.read_csv(('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_products_dataset.csv'))
sellers_df= pd.read_csv(('/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_sellers_dataset.csv'))
category_translation_df= pd.read_csv(('/content/drive/Othercomputers/My Laptop/Documents/Final test data/product_category_name_translation.csv'))


FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/Othercomputers/My Laptop/Documents/Final test data/olist_customers_dataset.csv'

In [ ]:
datasets = [customers_df, geolocation_df, items_df, payments_df, reviews_df, orders_df, products_df, sellers_df, category_translation_df]
titles = ["customers", "geolocation", "items", "payments", "reviews", "orders", "products", "sellers", "category_translation"]

data_summary = pd.DataFrame({},)
data_summary['datasets']= titles
data_summary['columns'] = [', '.join([col for col, null in data.isnull().sum().items() ]) for data in datasets]
data_summary['total_rows']= [data.shape[0] for data in datasets]
data_summary['total_cols']= [data.shape[1] for data in datasets]
data_summary['total_duplicate']= [len(data[data.duplicated()]) for data in datasets]
data_summary['total_null']= [data.isnull().sum().sum() for data in datasets]
data_summary['null_cols'] = [', '.join([col for col, null in data.isnull().sum().items() if null > 0]) for data in datasets]
for i in datasets:
    i.info()
    print("*"*50)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
**************************************************
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 

In [ ]:
for i in datasets:
    i.dropna(inplace=True)
for i in datasets:
    i.drop(i[i.duplicated()].index, axis=0, inplace=True)

In [ ]:
data_summary = pd.DataFrame({},)
data_summary['datasets']= titles
data_summary['columns'] = [', '.join([col for col, null in data.isnull().sum().items() ]) for data in datasets]
data_summary['total_rows']= [data.shape[0] for data in datasets]
data_summary['total_cols']= [data.shape[1] for data in datasets]
data_summary['total_duplicate']= [len(data[data.duplicated()]) for data in datasets]
data_summary['total_null']= [data.isnull().sum().sum() for data in datasets]
data_summary['null_cols'] = [', '.join([col for col, null in data.isnull().sum().items() if null > 0]) for data in datasets]
for i in datasets:
    i.info()
    print("*"*50)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB
**************************************************
<class 'pandas.core.frame.DataFrame'>
Int64Index: 738332 entries, 0 to 1000161
Data columns (total 5 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   geolocation_zip_code_prefix  738332 non-null  int64  
 1   geolocation_lat              738332 non-null  float64
 2   geolocation_lng              738332 non-null  float64
 3   ge

In [ ]:
merged_df= pd.merge(customers_df, orders_df, on="customer_id")
merged_df= merged_df.merge(reviews_df, on="order_id")
merged_df= merged_df.merge(items_df, on="order_id")
merged_df= merged_df.merge(products_df, on="product_id")
merged_df= merged_df.merge(payments_df, on="order_id")
merged_df= merged_df.merge(sellers_df, on='seller_id')
merged_df= merged_df.merge(category_translation_df, on='product_category_name')
merged_df.shape

(11578, 40)

In [ ]:
merged_df

customer_id                customer_unique_id  \
0      4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
1      ed9d62c3af1f6b24e984d012f66eca35  6a8fdf11c6cc553fe201f827da2f4e65   
2      9dca8837b7f9fc25ae379a58bb66094d  c207b575c7935606edefd74a43419a21   
3      e19274edf7a9c51689822810be84c9e9  854c37b10f4e835f6bc91355c9a7346c   
4      3d54e262ab34438486bd0eb8bf022c2f  f2479f8d117c49610d5a2d006b4ab716   
...                                 ...                               ...   
11573  806be13e55f43711b59f751f9ee83022  7eeb6421712844795dd8ff9ec6a58b18   
11574  806be13e55f43711b59f751f9ee83022  7eeb6421712844795dd8ff9ec6a58b18   
11575  806be13e55f43711b59f751f9ee83022  7eeb6421712844795dd8ff9ec6a58b18   
11576  806be13e55f43711b59f751f9ee83022  7eeb6421712844795dd8ff9ec6a58b18   
11577  aee0e70971aa92341d9c81d57319d6ee  ba1f5988dec5d5dcc36f2d14f2e94b3d   

       customer_zip_code_prefix       customer_city customer_state  \
0                         13056            campinas             SP   
1                         45970             itanhem             BA   
2                          6317         carapicuiba             SP   
3                          5772           sao paulo             SP   
4                         91240        porto alegre             RS   
...                         ...                 ...            ...   
11573                     33823  ribeirao das neves             MG   
11574                     33823  ribeirao das neves             MG   
11575                     33823  ribeirao das neves             MG   
11576                     33823  ribeirao das neves             MG   
11577                      4863           sao paulo             SP   

                               order_id order_status order_purchase_timestamp  \
0      6b7d50bd145f6fc7f33cebabd7e49d0f    delivered      2018-07-29 09:51:30   
1      1dfe970583e8a1a09241aa32a815a5b4    delivered      2018-07-18 17:21:49   
2      27fee5c620c806ed4f239ed553fca699    delivered      2018-05-22 12:27:08   
3      6f6785dde266434130a2aa8af06a52b7    delivered      2018-06-05 12:14:43   
4      00f251b943aba645652c7138e4716fee    delivered      2018-08-13 11:18:25   
...                                 ...          ...                      ...   
11573  6232b520aa859c47ba46e5213737f418    delivered      2018-06-15 12:52:05   
11574  6232b520aa859c47ba46e5213737f418    delivered      2018-06-15 12:52:05   
11575  6232b520aa859c47ba46e5213737f418    delivered      2018-06-15 12:52:05   
11576  6232b520aa859c47ba46e5213737f418    delivered      2018-06-15 12:52:05   
11577  cf439d61252ef55f81ce2e531d64a275    delivered      2018-05-15 16:45:05   

         order_approved_at order_delivered_carrier_date  ...  \
0      2018-07-29 10:10:09          2018-07-30 15:16:00  ...   
1      2018-07-18 17:35:14          2018-07-19 14:43:00  ...   
2      2018-05-22 12:56:01          2018-05-22 15:36:00  ...   
3      2018-06-05 14:31:37          2018-06-07 08:29:00  ...   
4      2018-08-13 11:44:26          2018-08-13 14:37:00  ...   
...                    ...                          ...  ...   
11573  2018-06-15 13:27:02          2018-06-19 13:53:00  ...   
11574  2018-06-15 13:27:02          2018-06-19 13:53:00  ...   
11575  2018-06-15 13:27:02          2018-06-19 13:53:00  ...   
11576  2018-06-15 13:27:02          2018-06-19 13:53:00  ...   
11577  2018-05-15 17:37:56          2018-05-16 13:49:00  ...   

      product_height_cm product_width_cm payment_sequential  payment_type  \
0                  15.0             35.0                  1   credit_card   
1                  15.0             35.0                  1   credit_card   
2                  10.0             40.0                  1   credit_card   
3                  10.0             40.0                  1   credit_card   
4                  10.0             40.0                  1   credit_card   
...                 ...              ...                ..

In [ ]:
merged_df.isnull().sum()

customer_id                      0
customer_unique_id               0
customer_zip_code_prefix         0
customer_city                    0
customer_state                   0
order_id                         0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
review_id                        0
review_score                     0
review_comment_title             0
review_comment_message           0
review_creation_date             0
review_answer_timestamp          0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
product_category_name            0
product_name_lenght              0
product_description_lenght       0
product_photos_qty               0
product_weight_g    

In [ ]:
len(merged_df[merged_df.duplicated()])

0

In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11578 entries, 0 to 11577
Data columns (total 40 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   customer_id                    11578 non-null  object 
 1   customer_unique_id             11578 non-null  object 
 2   customer_zip_code_prefix       11578 non-null  int64  
 3   customer_city                  11578 non-null  object 
 4   customer_state                 11578 non-null  object 
 5   order_id                       11578 non-null  object 
 6   order_status                   11578 non-null  object 
 7   order_purchase_timestamp       11578 non-null  object 
 8   order_approved_at              11578 non-null  object 
 9   order_delivered_carrier_date   11578 non-null  object 
 10  order_delivered_customer_date  11578 non-null  object 
 11  order_estimated_delivery_date  11578 non-null  object 
 12  review_id                      11578 non-null 

In [ ]:
time_columns= ['order_purchase_timestamp', 'order_approved_at','order_delivered_carrier_date','order_delivered_customer_date',
               'order_estimated_delivery_date', 'review_creation_date', 'review_answer_timestamp', 'shipping_limit_date']
merged_df[time_columns]=merged_df[time_columns].apply(pd.to_datetime)

In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11578 entries, 0 to 11577
Data columns (total 40 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   customer_id                    11578 non-null  object        
 1   customer_unique_id             11578 non-null  object        
 2   customer_zip_code_prefix       11578 non-null  int64         
 3   customer_city                  11578 non-null  object        
 4   customer_state                 11578 non-null  object        
 5   order_id                       11578 non-null  object        
 6   order_status                   11578 non-null  object        
 7   order_purchase_timestamp       11578 non-null  datetime64[ns]
 8   order_approved_at              11578 non-null  datetime64[ns]
 9   order_delivered_carrier_date   11578 non-null  datetime64[ns]
 10  order_delivered_customer_date  11578 non-null  datetime64[ns]
 11  order_estimated

In [ ]:
# RFM
import datetime as dt
present_day = merged_df['order_purchase_timestamp'].max() + dt.timedelta(days=2)
present_day

Timestamp('2018-08-31 14:18:28')

In [ ]:
print("Latest date in dataset: ", merged_df['order_purchase_timestamp'].max())

Latest date in dataset:  2018-08-29 14:18:28


In [ ]:
# Tìm R
recency_df= pd.DataFrame(merged_df.groupby(by='customer_unique_id', as_index=False)['order_purchase_timestamp'].max())
recency_df

customer_unique_id order_purchase_timestamp
0     0000366f3b9a7992bf8c76cfdf3221e2      2018-05-10 10:56:27
1     000ec5bff359e1c0ad76a81a45cb598f      2018-08-21 11:34:26
2     00172711b30d52eea8b313a7f2cced02      2018-07-28 00:23:49
3     001928b561575b2821c92254a2327d06      2018-08-24 21:17:00
4     0019e8c501c85848ac0966d45226fa1d      2018-05-01 09:25:32
...                                ...                      ...
9328  ffeb904468642a1ce663a322629801cb      2018-06-27 11:57:41
9329  ffecceca389973ef16660d58696f281e      2018-04-25 12:08:11
9330  fff2ae16b99c6f3c785f0e052f2a9cfb      2018-04-20 11:03:47
9331  fff3e1d7bc75f11dc7670619b2e61840      2018-07-20 13:47:30
9332  ffff5962728ec6157033ef9805bacc48      2018-05-02 15:17:41

[9333 rows x 2 columns]

In [ ]:
recency_df['Recency']= recency_df['order_purchase_timestamp'].apply(lambda x: (present_day - x).days)
recency_df

customer_unique_id order_purchase_timestamp  Recency
0     0000366f3b9a7992bf8c76cfdf3221e2      2018-05-10 10:56:27      113
1     000ec5bff359e1c0ad76a81a45cb598f      2018-08-21 11:34:26       10
2     00172711b30d52eea8b313a7f2cced02      2018-07-28 00:23:49       34
3     001928b561575b2821c92254a2327d06      2018-08-24 21:17:00        6
4     0019e8c501c85848ac0966d45226fa1d      2018-05-01 09:25:32      122
...                                ...                      ...      ...
9328  ffeb904468642a1ce663a322629801cb      2018-06-27 11:57:41       65
9329  ffecceca389973ef16660d58696f281e      2018-04-25 12:08:11      128
9330  fff2ae16b99c6f3c785f0e052f2a9cfb      2018-04-20 11:03:47      133
9331  fff3e1d7bc75f11dc7670619b2e61840      2018-07-20 13:47:30       42
9332  ffff5962728ec6157033ef9805bacc48      2018-05-02 15:17:41      120

[9333 rows x 3 columns]

In [ ]:
# Tìm F
frequency_df = pd.DataFrame(merged_df.groupby(["customer_unique_id"]).agg({"order_id":"nunique"}).reset_index())
frequency_df.rename(columns={"order_id":"Frequency"}, inplace=True)
frequency_df

customer_unique_id  Frequency
0     0000366f3b9a7992bf8c76cfdf3221e2          1
1     000ec5bff359e1c0ad76a81a45cb598f          1
2     00172711b30d52eea8b313a7f2cced02          1
3     001928b561575b2821c92254a2327d06          1
4     0019e8c501c85848ac0966d45226fa1d          1
...                                ...        ...
9328  ffeb904468642a1ce663a322629801cb          1
9329  ffecceca389973ef16660d58696f281e          1
9330  fff2ae16b99c6f3c785f0e052f2a9cfb          1
9331  fff3e1d7bc75f11dc7670619b2e61840          1
9332  ffff5962728ec6157033ef9805bacc48          1

[9333 rows x 2 columns]

In [ ]:
# Tìm M
monetary_df = merged_df.groupby('customer_unique_id', as_index=False)['payment_value'].sum()
monetary_df.columns = ['customer_unique_id', 'Monetary']
monetary_df.head()

customer_unique_id  Monetary
0  0000366f3b9a7992bf8c76cfdf3221e2    141.90
1  000ec5bff359e1c0ad76a81a45cb598f     27.75
2  00172711b30d52eea8b313a7f2cced02    122.07
3  001928b561575b2821c92254a2327d06    329.62
4  0019e8c501c85848ac0966d45226fa1d    197.27

In [ ]:
RF_df = recency_df.merge(frequency_df, on='customer_unique_id')
RFM_df = RF_df.merge(monetary_df, on='customer_unique_id').drop(columns='order_purchase_timestamp')
RFM_df.head()

customer_unique_id  Recency  Frequency  Monetary
0  0000366f3b9a7992bf8c76cfdf3221e2      113          1    141.90
1  000ec5bff359e1c0ad76a81a45cb598f       10          1     27.75
2  00172711b30d52eea8b313a7f2cced02       34          1    122.07
3  001928b561575b2821c92254a2327d06        6          1    329.62
4  0019e8c501c85848ac0966d45226fa1d      122          1    197.27

In [ ]:
RFM_df.describe().T

count        mean         std    min    25%     50%     75%  \
Recency    9333.0   74.851709   42.532132   2.00  37.00   75.00  112.00   
Frequency  9333.0    1.011036    0.108502   1.00   1.00    1.00    1.00   
Monetary   9333.0  245.638111  641.607276  13.89  68.18  122.42  217.55   

                max  
Recency      483.00  
Frequency      3.00  
Monetary   29099.52

In [ ]:
RFM_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9333 entries, 0 to 9332
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   customer_unique_id  9333 non-null   object 
 1   Recency             9333 non-null   int64  
 2   Frequency           9333 non-null   int64  
 3   Monetary            9333 non-null   float64
dtypes: float64(1), int64(2), object(1)
memory usage: 364.6+ KB


In [ ]:
# Kiểm tra các giá trị outlier
def num_outlier(df_in, col_name):
    q1 = df_in[col_name].quantile(0.05)
    q3 = df_in[col_name].quantile(0.95)
    iqr = q3-q1 #Interquartile range
    fence_low  = q1-1.5*iqr
    fence_high = q3+1.5*iqr
    outliers_df= df_in.loc[(df_in[col_name] < fence_low) | (df_in[col_name] > fence_high)]
    return print("Number of outliers in {} column: ".format(col_name), len(outliers_df)), print("Indexes: ", outliers_df.index)

In [ ]:
for i in ["Recency", "Frequency", "Monetary"]:
    num_outlier(RFM_df, i)
    print("*"*40)

Number of outliers in Recency column:  3
Indexes:  Int64Index([828, 2463, 2476], dtype='int64')
****************************************
Number of outliers in Frequency column:  99
Indexes:  Int64Index([  40,   96,  110,  138,  185,  238,  332,  335,  373,  508,  551,
             703,  874, 1093, 1234, 1333, 1428, 1500, 1739, 1786, 2146, 2156,
            2312, 2397, 2402, 2829, 2852, 2899, 2925, 2971, 3084, 3117, 3166,
            3668, 3740, 3824, 3849, 3928, 3938, 4102, 4113, 4221, 4235, 4292,
            4332, 4412, 4442, 4524, 4590, 4681, 4689, 4858, 4912, 5079, 5382,
            5412, 5500, 5672, 5763, 5900, 5945, 6005, 6083, 6090, 6108, 6172,
            6198, 6274, 6389, 6455, 6591, 6694, 6774, 6792, 7160, 7222, 7541,
            7590, 7651, 7763, 7807, 7982, 8085, 8151, 8257, 8417, 8428, 8609,
            8719, 8768, 8786, 8864, 8906, 8951, 8996, 9074, 9185, 9186, 9279],
           dtype='int64')
****************************************
Number of outliers in Monetary column: 

In [ ]:
RFM_df.shape

(9333, 4)

In [ ]:
# Loại bỏ các giá trị outlier
def remove_outlier(df_in, col_name):
    q1 = df_in[col_name].quantile(0.05)
    q3 = df_in[col_name].quantile(0.95)
    iqr = q3-q1
    fence_low  = q1-1.5*iqr
    fence_high = q3+1.5*iqr
    index_outliers= df_in.loc[(df_in[col_name] < fence_low) | (df_in[col_name] > fence_high)].index
    df_in= pd.DataFrame(df_in.drop(index_outliers.to_list(), axis=0, inplace=True))
    print("Outliers in the {} column have been removed".format(col_name))
    return df_in

In [ ]:
for i in ["Recency", "Monetary"]:
    remove_outlier(RFM_df, i)
    print("*"*40)

Outliers in the Recency column have been removed
****************************************
Outliers in the Monetary column have been removed
****************************************


In [ ]:
RFM_df.shape

(9230, 4)

In [ ]:
RFM_df2= RFM_df.copy()
RFM_df2= RFM_df2.set_index('customer_unique_id')
RFM_df2

Recency  Frequency  Monetary
customer_unique_id                                            
0000366f3b9a7992bf8c76cfdf3221e2      113          1    141.90
000ec5bff359e1c0ad76a81a45cb598f       10          1     27.75
00172711b30d52eea8b313a7f2cced02       34          1    122.07
001928b561575b2821c92254a2327d06        6          1    329.62
0019e8c501c85848ac0966d45226fa1d      122          1    197.27
...                                   ...        ...       ...
ffeb904468642a1ce663a322629801cb       65          1     93.11
ffecceca389973ef16660d58696f281e      128          1     72.76
fff2ae16b99c6f3c785f0e052f2a9cfb      133          1    200.90
fff3e1d7bc75f11dc7670619b2e61840       42          1     82.51
ffff5962728ec6157033ef9805bacc48      120          1    133.69

[9230 rows x 3 columns]

In [ ]:
# Creating RFM Segments
RFM_df2["recency_score"]  = pd.qcut(RFM_df2['Recency'], 5, labels=[5, 4, 3, 2, 1])
RFM_df2["frequency_score"]= pd.qcut(RFM_df2['Frequency'].rank(method="first"), 5, labels=[1, 2, 3, 4, 5])
RFM_df2["monetary_score"] = pd.qcut(RFM_df2['Monetary'], 5, labels=[1, 2, 3, 4, 5])
RFM_df2['RFM_SCORE'] = RFM_df2.recency_score.astype(str)+ RFM_df2.frequency_score.astype(str) + RFM_df2.monetary_score.astype(str)
RFM_df2

Recency  Frequency  Monetary recency_score  \
customer_unique_id                                                             
0000366f3b9a7992bf8c76cfdf3221e2      113          1    141.90             2   
000ec5bff359e1c0ad76a81a45cb598f       10          1     27.75             5   
00172711b30d52eea8b313a7f2cced02       34          1    122.07             4   
001928b561575b2821c92254a2327d06        6          1    329.62             5   
0019e8c501c85848ac0966d45226fa1d      122          1    197.27             1   
...                                   ...        ...       ...           ...   
ffeb904468642a1ce663a322629801cb       65          1     93.11             3   
ffecceca389973ef16660d58696f281e      128          1     72.76             1   
fff2ae16b99c6f3c785f0e052f2a9cfb      133          1    200.90             1   
fff3e1d7bc75f11dc7670619b2e61840       42          1     82.51             4   
ffff5962728ec6157033ef9805bacc48      120          1    133.69             1   

                                 frequency_score monetary_score RFM_SCORE  
customer_unique_id                                                         
0000366f3b9a7992bf8c76cfdf3221e2               1              3       213  
000ec5bff359e1c0ad76a81a45cb598f               1              1       511  
00172711b30d52eea8b313a7f2cced02               1              3       413  
001928b561575b2821c92254a2327d06               1              5       515  
0019e8c501c85848ac0966d45226fa1d               1              4       114  
...                                          ...            ...       ...  
ffeb904468642a1ce663a322629801cb               5              2       352  
ffecceca389973ef16660d58696f281e               5              2       152  
fff2ae16b99c6f3c785f0e052f2a9cfb               5              4       154  
fff3e1d7bc75f11dc7670619b2e61840               5              2       452  
ffff5962728ec6157033ef9805bacc48               5              3       153  

[9230 rows x 7 columns]

In [ ]:
seg_map= {
    r'111|112|121|131|141|151': 'Lost customers',
    r'332|322|233|232|223|222|132|123|122|212|211': 'Hibernating customers',
    r'155|154|144|214|215|115|114|113': 'Cannot Lose Them',
    r'255|254|245|244|253|252|243|242|235|234|225|224|153|152|145|143|142|135|134|133|125|124': 'At Risk',
    r'331|321|312|221|213|231|241|251': 'About To Sleep',
    r'535|534|443|434|343|334|325|324': 'Need Attention',
    r'525|524|523|522|521|515|514|513|425|424|413|414|415|315|314|313': 'Promising',
    r'512|511|422|421|412|411|311': 'New Customers',
    r'553|551|552|541|542|533|532|531|452|451|442|441|431|453|433|432|423|353|352|351|342|341|333|323': 'Potential Loyalist',
    r'543|444|435|355|354|345|344|335': 'Loyal',
    r'555|554|544|545|454|455|445': 'Champions'
}

In [ ]:
RFM_df2['Segment'] = RFM_df2['recency_score'].astype(str) + RFM_df2['frequency_score'].astype(str) + RFM_df2['monetary_score'].astype(str)
RFM_df2['Segment'] = RFM_df2['Segment'].replace(seg_map, regex=True)
RFM_df2.head()

Recency  Frequency  Monetary recency_score  \
customer_unique_id                                                             
0000366f3b9a7992bf8c76cfdf3221e2      113          1    141.90             2   
000ec5bff359e1c0ad76a81a45cb598f       10          1     27.75             5   
00172711b30d52eea8b313a7f2cced02       34          1    122.07             4   
001928b561575b2821c92254a2327d06        6          1    329.62             5   
0019e8c501c85848ac0966d45226fa1d      122          1    197.27             1   

                                 frequency_score monetary_score RFM_SCORE  \
customer_unique_id                                                          
0000366f3b9a7992bf8c76cfdf3221e2               1              3       213   
000ec5bff359e1c0ad76a81a45cb598f               1              1       511   
00172711b30d52eea8b313a7f2cced02               1              3       413   
001928b561575b2821c92254a2327d06               1              5       515   
0019e8c501c85848ac0966d45226fa1d               1              4       114   

                                           Segment  
customer_unique_id                                  
0000366f3b9a7992bf8c76cfdf3221e2    About To Sleep  
000ec5bff359e1c0ad76a81a45cb598f     New Customers  
00172711b30d52eea8b313a7f2cced02         Promising  
001928b561575b2821c92254a2327d06         Promising  
0019e8c501c85848ac0966d45226fa1d  Cannot Lose Them

In [ ]:
RFM_df2['Segment'].unique()

array(['About To Sleep', 'New Customers', 'Promising', 'Cannot Lose Them',
       'Hibernating customers', 'Lost customers', 'Champions', 'At Risk',
       'Potential Loyalist', 'Loyal', 'Need Attention'], dtype=object)

In [ ]:
RFM_df2['Segment'].nunique()

11

In [ ]:
RFMStats = RFM_df2[["Segment", "Recency", "Frequency", "Monetary"]].groupby("Segment").agg(['mean','median', 'min', 'max', 'count'])
RFMStats

Recency                        Frequency             \
                             mean median  min  max count      mean median min   
Segment                                                                         
About To Sleep          93.025000   96.0   60  117   600  1.001667    1.0   1   
At Risk                117.980271  116.0   89  315  1622  1.012330    1.0   1   
Cannot Lose Them       125.384740  125.0   89  272   616  1.009740    1.0   1   
Champions               30.048682   29.0    2   59   493  1.091278    1.0   1   
Hibernating customers  107.546735  108.0   60  311   781  1.000000    1.0   1   
Lost customers         132.042289  129.0  118  261   402  1.000000    1.0   1   
Loyal                   61.446918   67.0    2   88   584  1.025685    1.0   1   
Need Attention          52.857645   58.0    2   88   569  1.000000    1.0   1   
New Customers           40.662857   38.0    2   88   525  1.000000    1.0   1   
Potential Loyalist      43.570724   40.0    2   88  1824  1.008224    1.0   1   
Promising               37.837727   31.0    2   88  1214  1.000000    1.0   1   

                                   Monetary                                  
                      max count        mean   median     min      max count  
Segment                                                                      
About To Sleep          2   600   56.353850   47.750   14.38   147.57   600  
At Risk                 3  1622  287.677275  175.015   58.63  1919.44  1622  
Cannot Lose Them        2   616  321.003961  209.260   98.43  1890.18   616  
Champions               3   493  404.530974  299.070  149.66  1868.97   493  
Hibernating customers   1   781   87.082330   85.300   16.29   148.94   781  
Lost customers          1   402   48.456219   47.225   17.29    97.38   402  
Loyal                   2   584  403.679452  258.910   97.67  1953.15   584  
Need Attention          1   569  255.796450  180.370   97.65  1679.68   569  
New Customers           1   525   56.816933   53.380   14.29    97.59   525  
Potential Loyalist      2  1824   79.567697   76.360   13.89   149.01  1824  
Promising               1  1214  266.907142  179.155   18.70  1957.20  1214

In [ ]:
RFMStats['Ratio']= (100*RFMStats['Monetary']["count"]/RFMStats['Monetary']["count"].sum()).round(2)
RFMStats

Recency                        Frequency             \
                             mean median  min  max count      mean median min   
Segment                                                                         
About To Sleep          93.025000   96.0   60  117   600  1.001667    1.0   1   
At Risk                117.980271  116.0   89  315  1622  1.012330    1.0   1   
Cannot Lose Them       125.384740  125.0   89  272   616  1.009740    1.0   1   
Champions               30.048682   29.0    2   59   493  1.091278    1.0   1   
Hibernating customers  107.546735  108.0   60  311   781  1.000000    1.0   1   
Lost customers         132.042289  129.0  118  261   402  1.000000    1.0   1   
Loyal                   61.446918   67.0    2   88   584  1.025685    1.0   1   
Need Attention          52.857645   58.0    2   88   569  1.000000    1.0   1   
New Customers           40.662857   38.0    2   88   525  1.000000    1.0   1   
Potential Loyalist      43.570724   40.0    2   88  1824  1.008224    1.0   1   
Promising               37.837727   31.0    2   88  1214  1.000000    1.0   1   

                                   Monetary                                  \
                      max count        mean   median     min      max count   
Segment                                                                       
About To Sleep          2   600   56.353850   47.750   14.38   147.57   600   
At Risk                 3  1622  287.677275  175.015   58.63  1919.44  1622   
Cannot Lose Them        2   616  321.003961  209.260   98.43  1890.18   616   
Champions               3   493  404.530974  299.070  149.66  1868.97   493   
Hibernating customers   1   781   87.082330   85.300   16.29   148.94   781   
Lost customers          1   402   48.456219   47.225   17.29    97.38   402   
Loyal                   2   584  403.679452  258.910   97.67  1953.15   584   
Need Attention          1   569  255.796450  180.370   97.65  1679.68   569   
New Customers           1   525   56.816933   53.380   14.29    97.59   525   
Potential Loyalist      2  1824   79.567697   76.360   13.89   149.01  1824   
Promising               1  1214  266.907142  179.155   18.70  1957.20  1214   

                       Ratio  
                              
Segment                       
About To Sleep          6.50  
At Risk                17.57  
Cannot Lose Them        6.67  
Champions               5.34  
Hibernating customers   8.46  
Lost customers          4.36  
Loyal                   6.33  
Need Attention          6.16  
New Customers           5.69  
Potential Loyalist     19.76  
Promising              13.15

1. Nhóm khách hàng Best Customers có đặc điểm như thế nào? Số lượng?
=> Khách hàng best customer thuộc segment Champions. Đây là những khách hàng thường xuyên mua sắm và chi tiêu nhiều nhất so với các khách hàng còn lại dựa trên bộ dữ liệu của Olist cung cấp.
=> Đặc điểm nhận dạng: Những khách hàng đã mua hàng cách đây từ 2 - 59 ngày. Với mỗi khách hàng đó trung bình họ sẽ ghé cửa hàng nhiều hơn 1 lần. Đồng thời, họ chi tiêu khoảng 404.530974. Số tiền ít nhất họ bỏ ra là 149.66 và cao nhất là 1868.97. (Đơn vị tính là tiền Brazil)
=> Tổng số lượng khách hàng Best Customers (Champions): 439

In [ ]:
RFM_df2[RFM_df2['Segment'].isin ({'Champions'})]

Recency  Frequency  Monetary recency_score  \
customer_unique_id                                                             
013ef03e0f3f408dd9bf555e4edcdc0a       42          2    245.37             4   
03ef7f03905227e15c7b6c860554b3cb       44          2    238.55             4   
1772c0fbee58b154da028487de51b74c       44          2    772.58             4   
240f980f7c03e873c538e86792646919       23          2    553.80             5   
30b782a79466007756f170cb5bd6bbd8       31          2    551.23             4   
...                                   ...        ...       ...           ...   
fefa1fa3498777f7cb5b4a21ee8b54a5       16          1    309.84             5   
ff0c3ce5dfd690ab85352fc85683303b       35          1    318.30             4   
ff176d08b6e19c92b5ccfcdfa01cdffa        9          1    170.14             5   
ff2959dedacc441cc5ffe7d747a1a299       13          1   1043.32             5   
ffe204b1bb8026741b839085b2fd10f2       21          1    156.70             5   

                                 frequency_score monetary_score RFM_SCORE  \
customer_unique_id                                                          
013ef03e0f3f408dd9bf555e4edcdc0a               5              4       454   
03ef7f03905227e15c7b6c860554b3cb               5              4       454   
1772c0fbee58b154da028487de51b74c               5              5       455   
240f980f7c03e873c538e86792646919               5              5       555   
30b782a79466007756f170cb5bd6bbd8               5              5       455   
...                                          ...            ...       ...   
fefa1fa3498777f7cb5b4a21ee8b54a5               5              5       555   
ff0c3ce5dfd690ab85352fc85683303b               5              5       455   
ff176d08b6e19c92b5ccfcdfa01cdffa               5              4       554   
ff2959dedacc441cc5ffe7d747a1a299               5              5       555   
ffe204b1bb8026741b839085b2fd10f2               5              4       554   

                                    Segment  
customer_unique_id                           
013ef03e0f3f408dd9bf555e4edcdc0a  Champions  
03ef7f03905227e15c7b6c860554b3cb  Champions  
1772c0fbee58b154da028487de51b74c  Champions  
240f980f7c03e873c538e86792646919  Champions  
30b782a79466007756f170cb5bd6bbd8  Champions  
...                                     ...  
fefa1fa3498777f7cb5b4a21ee8b54a5  Champions  
ff0c3ce5dfd690ab85352fc85683303b  Champions  
ff176d08b6e19c92b5ccfcdfa01cdffa  Champions  
ff2959dedacc441cc5ffe7d747a1a299  Champions  
ffe204b1bb8026741b839085b2fd10f2  Champions  

[493 rows x 8 columns]

2. Đâu là các khách hàng rời bỏ?  Số lượng? (1.5đ)
=> Theo mô hình RFM, nhóm khách hàng rời bỏ, còn được gọi là “At Risk Customers”, là những khách hàng đã từng mua hàng nhưng không mua hàng gần đây.
=> Như vậy khách hàng rời bỏ thuộc segment At Risk.
=> Đặc điểm nhận dạng: Những khách mua hàng cách đây 89- 315 ngày. Họ ghé cửa hàng ít nhất hơn 1 lần (từ 1 đến 3 lần). Trung bình họ sẽ chi tiêu khoảng 287.677275, với số tiền ít nhất là 58.63 và nhiều nhất là 1919.44 (Tiền Brazil).

In [ ]:
RFM_df2[RFM_df2['Segment'].isin ({'At Risk'})]

Recency  Frequency  Monetary recency_score  \
customer_unique_id                                                             
02e9109b7e0a985108b43e573b6afb23      109          2    598.81             2   
08d8ad84b0088cbeae77da4ff3817479      106          2    259.81             2   
08e5b38d7948d37fbb2a59fc5e175ab1      113          3    701.43             2   
09cd73b13f103e47ce13675b4a4b36f3      112          2    375.40             2   
0e850ed5adc4abe0346435eda5b1a626      140          2     65.06             1   
...                                   ...        ...       ...           ...   
ffb1cf5736b51e47fec82cf40dd26303       96          1    267.66             2   
ffb7a3f4d1afe09714753add843d23b0      121          1    136.71             1   
ffd6f65402f2bc47238ecd2bdc93e0d4      113          1    166.92             2   
ffecceca389973ef16660d58696f281e      128          1     72.76             1   
ffff5962728ec6157033ef9805bacc48      120          1    133.69             1   

                                 frequency_score monetary_score RFM_SCORE  \
customer_unique_id                                                          
02e9109b7e0a985108b43e573b6afb23               5              5       255   
08d8ad84b0088cbeae77da4ff3817479               5              5       255   
08e5b38d7948d37fbb2a59fc5e175ab1               5              5       255   
09cd73b13f103e47ce13675b4a4b36f3               5              5       255   
0e850ed5adc4abe0346435eda5b1a626               5              2       152   
...                                          ...            ...       ...   
ffb1cf5736b51e47fec82cf40dd26303               5              5       255   
ffb7a3f4d1afe09714753add843d23b0               5              3       153   
ffd6f65402f2bc47238ecd2bdc93e0d4               5              4       254   
ffecceca389973ef16660d58696f281e               5              2       152   
ffff5962728ec6157033ef9805bacc48               5              3       153   

                                  Segment  
customer_unique_id                         
02e9109b7e0a985108b43e573b6afb23  At Risk  
08d8ad84b0088cbeae77da4ff3817479  At Risk  
08e5b38d7948d37fbb2a59fc5e175ab1  At Risk  
09cd73b13f103e47ce13675b4a4b36f3  At Risk  
0e850ed5adc4abe0346435eda5b1a626  At Risk  
...                                   ...  
ffb1cf5736b51e47fec82cf40dd26303  At Risk  
ffb7a3f4d1afe09714753add843d23b0  At Risk  
ffd6f65402f2bc47238ecd2bdc93e0d4  At Risk  
ffecceca389973ef16660d58696f281e  At Risk  
ffff5962728ec6157033ef9805bacc48  At Risk  

[1622 rows x 8 columns]

3. Nhóm khách hàng có khả năng Churned có đặc điểm như thế nào ?  (1.5đ)
Trong các nhóm khách hàng đã liệt kê theo mô hình RFM, nhóm khách hàng “Churned” có thể bao gồm các nhóm sau:
‘About To Sleep’: Những khách hàng này đã không mua hàng trong một thời gian dài, có thể đang chuẩn bị rời bỏ.
‘Hibernating customers’: Những khách hàng này đã không mua hàng trong một thời gian rất dài, có thể đã rời bỏ.
‘Lost customers’: Những khách hàng này đã không mua hàng trong một thời gian rất dài và có lượng mua hàng thấp, rất có thể đã rời bỏ.
‘At Risk’: Những khách hàng này đã từng mua hàng nhưng không mua hàng gần đây, có nguy cơ rất cao rời bỏ.

=> Đặc điểm nhận dạng: Họ là những ngày mua hàng gần nhất cách ngày hiện tại từ 60 đến 315 ngày. Mức độ thường xuyên lui đến cửa hàng ít nhất 1 lần với tổng chi tiêu trung bình là 172.662423. Số tiền lớn nhất mà họ chi tiêu là 1919.44 và thấp nhất là 14.38 (Tiền Brazil).


In [ ]:
RFM_df2['Segment'] = RFM_df2['Segment'].replace(['About To Sleep', 'Hibernating customers', 'Lost customers', 'At Risk'], 'About To Sleep & Hibernating customers & Lost customers & At Risk')
RFMStats_2 = RFM_df2[["Segment", "Recency", "Frequency", "Monetary"]].groupby('Segment').agg(['mean','median', 'min', 'max', 'count'])
RFMStats_2

Recency             \
                                                          mean median min   
Segment                                                                     
About To Sleep & Hibernating customers & Lost c...  112.849927  113.0  60   
Cannot Lose Them                                    125.384740  125.0  89   
Champions                                            30.048682   29.0   2   
Loyal                                                61.446918   67.0   2   
Need Attention                                       52.857645   58.0   2   
New Customers                                        40.662857   38.0   2   
Potential Loyalist                                   43.570724   40.0   2   
Promising                                            37.837727   31.0   2   

                                                              Frequency  \
                                                    max count      mean   
Segment                                                                   
About To Sleep & Hibernating customers & Lost c...  315  3405  1.006167   
Cannot Lose Them                                    272   616  1.009740   
Champions                                            59   493  1.091278   
Loyal                                                88   584  1.025685   
Need Attention                                       88   569  1.000000   
New Customers                                        88   525  1.000000   
Potential Loyalist                                   88  1824  1.008224   
Promising                                            88  1214  1.000000   

                                                                         \
                                                   median min max count   
Segment                                                                   
About To Sleep & Hibernating customers & Lost c...    1.0   1   3  3405   
Cannot Lose Them                                      1.0   1   2   616   
Champions                                             1.0   1   3   493   
Loyal                                                 1.0   1   2   584   
Need Attention                                        1.0   1   1   569   
New Customers                                         1.0   1   1   525   
Potential Loyalist                                    1.0   1   2  1824   
Promising                                             1.0   1   1  1214   

                                                      Monetary           \
                                                          mean   median   
Segment                                                                   
About To Sleep & Hibernating customers & Lost c...  172.662423   95.260   
Cannot Lose Them                                    321.003961  209.260   
Champions                                           404.530974  299.070   
Loyal                                               403.679452  258.910   
Need Attention                                      255.796450  180.370   
New Customers                                        56.816933   53.380   
Potential Loyalist                                   79.567697   76.360   
Promising                                           266.907142  179.155   

                                                                           
                                                       min      max count  
Segment                                                                    
About To Sleep & Hibernating customers & Lost c...   14.38  1919.44  3405  
Cannot Lose Them                                     98.43  1890.18   616  
Champions                                           149.66  1868.97   493  
Loyal                                                97.67  1953.15   584  
Need Attention                                       97.65  1679.68   569  
New Customers                                        14.29    97.59   525  
Potential Loyalist                           

In [ ]:
RFM_df2[RFM_df2['Segment'].isin ({'About To Sleep', 'Hibernating customers', 'Lost customers', 'At Risk'})]

Empty DataFrame
Columns: [Recency, Frequency, Monetary, recency_score, frequency_score, monetary_score, RFM_SCORE, Segment]
Index: []